### PR TITLE
Adapt dubbo protocol with Istio httproute

### PR DESCRIPTION
**Pull request by champly**
### Issues associated with this PR

new pr with #1122 

### Solutions

Use Istio VirtualSerivce httproute adapte dubbo protocol

### UT result
Unit Test is needed if the code is changed, your unit test should cover boundary cases, corner cases, and some exceptional cases. And you need to show the UT result.

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result.

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

File: pkg/filter/stream/dubbo/dubbo.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+func (f *factory) CreateFilterChain(ctx context.Context, callbacks api.StreamFilterChainFactoryCallbacks) {
+	filter := NewDubboFilter(ctx)
+	callbacks.AddStreamReceiverFilter(filter, api.BeforeRoute)
+	callbacks.AddStreamSenderFilter(filter)
+}
+
+type dubboFilter struct {
+	handler api.StreamReceiverFilterHandler
```
Comment (cch123) [inline pkg/filter/stream/dubbo/dubbo.go:45]: check about this warning

File: pkg/filter/stream/dubbo/install/build/mosn_config_dubbo_xds.json [hunk 2]
```diff
@@ -0,0 +1,189 @@
+{
```
Comment (cch123) [inline pkg/filter/stream/dubbo/install/build/mosn_config_dubbo_xds.json:1]: is it properer to put these config files and install scripts to example dir

File: pkg/filter/stream/dubbo/dubbo.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+func (d *dubboFilter) OnDestroy() {}
+
+func (d *dubboFilter) OnReceive(ctx context.Context, headers api.HeaderMap, buf buffer.IoBuffer, trailers api.HeaderMap) api.StreamFilterStatus {
+
+	listener := mosnctx.Get(ctx, types.ContextKeyListenerName).(string)
+
+	service, ok := headers.Get(dubbo.ServiceNameHeader)
+	if ok {
```
Comment (cch123) [inline pkg/filter/stream/dubbo/dubbo.go:59]: what will happen when it's not ok?

File: pkg/filter/stream/dubbo/dubbo.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+	return &dubboFilter{}
+}
+
+func (d *dubboFilter) OnDestroy() {}
+
+func (d *dubboFilter) OnReceive(ctx context.Context, headers api.HeaderMap, buf buffer.IoBuffer, trailers api.HeaderMap) api.StreamFilterStatus {
+
+	listener := mosnctx.Get(ctx, types.ContextKeyListenerName).(string)
+
+	service, ok := headers.Get(dubbo.ServiceNameHeader)
```
Comment (cch123) [inline pkg/filter/stream/dubbo/dubbo.go:58]: what will happen here when ok == false

File: pkg/filter/stream/dubbo/dubbo.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+	listener := mosnctx.Get(ctx, types.ContextKeyListenerName).(string)
+	service := mosnctx.Get(ctx, types.ContextKeyRouteService).(string)
+	method := mosnctx.Get(ctx, types.ContextKeyRouteMethod).(string)
+
+	stats := getStats(listener, service, method)
+	if stats == nil {
+		return api.StreamFilterContinue
+	}
+
+	if frame.GetStatusCode() == 20 {
```
Comment (cch123) [inline pkg/filter/stream/dubbo/dubbo.go:101]: 20 is a magic number, please define a constant, and it's better to link to a dubbo official doc

**Comment by cch123** [comment]
LGTM

File: pkg/filter/stream/dubbo/dubbo.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+func (d *dubboFilter) OnDestroy() {}
+
+func (d *dubboFilter) OnReceive(ctx context.Context, headers api.HeaderMap, buf buffer.IoBuffer, trailers api.HeaderMap) api.StreamFilterStatus {
+
+	listener := mosnctx.Get(ctx, types.ContextKeyListenerName).(string)
+
+	service, ok := headers.Get(dubbo.ServiceNameHeader)
+	if !ok {
+		log.DefaultLogger.Errorf("This filter {%s} just for dubbo protocol, please check your config.", v2.DubboStream)
+		return api.StreamFilterStop
```
Comment (nejisama) [inline pkg/filter/stream/dubbo/dubbo.go:68]: if OnReceive returns StreamFilterStop , it means stop next stream filters, but do not stop the request.
Stop request should returns StreamFiltertermination. 
Just ingore the filter should  returns Continue, run the other filters.

Comment (champly) [reply pkg/filter/stream/dubbo/dubbo.go:68]: It should return StreamFiltertermination

**Comment by champly** [comment]
- Retrieving the data of service & instances from a registry center such as Nacos, zookeeper.
- Generating the istio's CRD through various data such as registry data & configurations.
- Make MOSN be able to support serve a registry request which comes from a dubbo provider.

architecture-diagram

**Comment by cch123** [comment]
LGTM

**Review by cch123 — APPROVED** [review]

File: pkg/filter/stream/dubbo/dubbo.go [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+	headers.Set(protocol.MosnHeaderHostKey, service)
+	// because use http rule, so should add default path
+	headers.Set(protocol.MosnHeaderPathKey, "/")
+
+	method, _ := headers.Get(dubbo.MethodNameHeader)
+	stats := getStats(listener, service, method)
+	if stats != nil {
+		stats.RequestServiceInfo.Inc(1)
+
+		variable.SetVariableValue(ctx, dubboStreamService, service)
```
Comment (taoyuanyuan) [inline pkg/filter/stream/dubbo/dubbo.go:81]: you should registry variable in buildStream

File: pkg/filter/stream/dubbo/stats.go [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+type Stats struct {
+	RequestServiceInfo gometrics.Counter
+	ResponseSucc       gometrics.Counter
+	ResponseFail       gometrics.Counter
+}
+
+func getStats(listener, service, method string) *Stats {
+	key := service + "-" + method
+	if s, ok := statsFactory[key]; ok {
```
Comment (taoyuanyuan) [inline pkg/filter/stream/dubbo/stats.go:39]: concurrent map read and map write

**Comment by taoyuanyuan** [comment]
Please run the example successfully after bugfix.

**Comment by champly** [comment]
> Please run the example successfully after bugfix.

Already registry variable and run example success.

mosn

File: pkg/filter/stream/dubbo/var.go [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+var (
+	buildinVariables = []variable.Variable{
+		variable.NewIndexedVariable(types.VarDubboRequestService, nil, nil, variable.BasicSetter, 0),
+		variable.NewIndexedVariable(types.VarDubboRequestMethod, nil, nil, variable.BasicSetter, 0),
+	}
+)
+
+func init() {
+	// variable must registry
+	for idx := range buildinVariables {
```
Comment (taoyuanyuan) [inline pkg/filter/stream/dubbo/var.go:17]: you should registry when dubbo filter enable.

Comment (taoyuanyuan) [reply pkg/filter/stream/dubbo/var.go:17]: run example again.

Comment (champly) [reply pkg/filter/stream/dubbo/var.go:17]: Alreay run example succ.

**Review by taoyuanyuan — APPROVED** [review]
LGTM
